File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        int l = 0; int r = n - 1;
        int maxRight = 0;
        int maxLeft = 0;
        int water = 0;
        while(l<r){
            maxLeft = max(height[l], maxLeft);
            maxRight = max(height[r], maxRight);
            if(maxLeft > maxRight){
                water += maxRight - height[r];
                r--;
            }
            else {
                water += maxLeft - height[l];
                l++;
            }
        }
        return water;
    }
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.go
func trap(height []int) int {
    sz := len(height)
    
    maxFromL := make([]int, sz)
    maxFromR := make([]int, sz)
    
    maxFromL[0] = height[0]
    maxFromR[sz-1] = height[sz-1]
    
    maxiL := maxFromL[0]
    maxiR := maxFromR[sz-1]
    
    totalWater := 0
    
    for i:= 1; i<sz; i++{
        maxFromL[i] = max(maxiL, height[i])
        maxiL = maxFromL[i] 
    }

    for i:=sz-2; i>=0; i-- {
        maxFromR[i] = max(maxiR, height[i])
        maxiR = maxFromR[i]
    }

    for i:=0;i<sz;i++{
        totalWater += min(maxFromL[i], maxFromR[i]) - height[i]
    }
    return totalWater
   
}